### PR TITLE
[ci/coverage][1] A script to collect dynamic test coverage

### PR DESCRIPTION
**Pull request by can-anyscale**
## Why are these changes needed?
A simple script to run the coverage tool to collect test coverage. Nothing fancy at this point, but I want this check in so I can test if can run this successfully in a buildkite environment.

## Checks

- [X] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [X] I've run `scripts/format.sh` to lint the changes in this PR.
- Testing Strategy

```
(base) can@can-Q7M1Q6FVFV ray % python ci/test/ray_coverage.py release/...                         
2023-05-03 15:10:50,830:INFO:root:Collecting coverage for test target: release/...
Loading: 
Loading: 0 packages loaded
DEBUG: /Users/can/Projects/ray/bazel/ray_deps_setup.bzl:67:14: No implicit mirrors used because urls were explicitly provided
DEBUG: Rule 'rules_perl' indicated that a canonical reproducible form can be obtained by modifying arguments shallow_since = "1663780239 -0700"
DEBUG: Repository rules_perl instantiated at:
  /Users/can/Projects/ray/WORKSPACE:6:15: in <toplevel>
  /Users/can/Projects/ray/bazel/ray_deps_setup.bzl:260:19: in ray_deps_setup
Repository rule git_repository defined at:
  /private/var/tmp/_bazel_can/bbbe52dc6476b8ea93cfccfdb9e74643/external/bazel_tools/tools/build_defs/repo/git.bzl:199:33: in <toplevel>
Analyzing: 29 targets (0 packages loaded, 0 targets configured)
INFO: Analyzed 29 targets (0 packages loaded, 0 targets configured).
INFO: Found 20 targets and 9 test targets...
[0 / 1] [Prepa] BazelWorkspaceStatusAction stable-status.txt
[2 / 3] 1 / 9 tests; Testing //release:test_anyscale_job_wrapper; 1s darwin-sandbox
[2 / 3] 1 / 9 tests; Testing //release:test_anyscale_job_wrapper; 11s darwin-sandbox
[2 / 3] 1 / 9 tests; Testing //release:test_anyscale_job_wrapper; 14s darwin-sandbox
[4 / 5] 3 / 9 tests; Testing //release:test_glue; 1s darwin-sandbox
[6 / 7] 5 / 9 tests; Testing //release:test_config; 2s darwin-sandbox
[9 / 10] 8 / 9 tests; Testing //release:test_wheels; 0s darwin-sandbox
INFO: Elapsed time: 24.616s, Critical Path: 14.84s
INFO: 10 processes: 18 darwin-sandbox.
INFO: Build completed successfully, 10 total actions
INFO: Build completed successfully, 10 total actions
2023-05-03 15:11:15,702:INFO:root:Name                                                          Stmts   Miss  Cover
---------------------------------------------------------------------------------
release/__init__.py                                               0      0   100%
release/ray_release/__init__.py                                   0      0   100%
release/ray_release/alerts/__init__.py                            0      0   100%
release/ray_release/alerts/default.py                             7      4    43%
release/ray_release/alerts/handle.py                             21      9    57%
release/ray_release/alerts/long_running_tests.py                 18     18     0%
release/ray_release/alerts/tune_tests.py                         44     44     0%
release/ray_release/alerts/xgboost_tests.py                      34     34     0%
release/ray_release/anyscale_util.py                             30     27    10%
release/ray_release/aws.py                                       45     34    24%
release/ray_release/buildkite/__init__.py                         0      0   100%
release/ray_release/buildkite/concurrency.py                     64     21    67%
release/ray_release/buildkite/filter.py                          45     11    76%
release/ray_release/buildkite/output.py                          13      7    46%
release/ray_release/buildkite/settings.py                       124     44    65%
release/ray_release/buildkite/step.py                            60     21    65%
release/ray_release/cluster_manager/__init__.py                   0      0   100%
release/ray_release/cluster_manager/cluster_manager.py           79     34    57%
release/ray_release/cluster_manager/full.py                      60     20    67%
release/ray_release/cluster_manager/minimal.py                  157     39    75%
release/ray_release/command_runner/__init__.py                    0      0   100%
release/ray_release/command_runner/_anyscale_job_wrapper.py     192     60    69%
release/ray_release/command_runner/_prometheus_metrics.py        76     76     0%
release/ray_release/command_runner/_wait_cluster.py              25     25     0%
release/ray_release/command_runner/anyscale_job_runner.py       150    150     0%
release/ray_release/command_runner/command_runner.py             51     41    20%
release/ray_release/command_runner/job_runner.py                 76     76     0%
release/ray_release/config.py                                   141     64    55%
release/ray_release/env.py                                       20      7    65%
release/ray_release/exception.py                                 85     85     0%
release/ray_release/file_manager/__init__.py                      0      0   100%
release/ray_release/file_manager/file_manager.py                 10      9    10%
release/ray_release/file_manager/job_file_manager.py            104     95     9%
release/ray_release/file_manager/remote_task.py                  42     42     0%
release/ray_release/glue.py                                     258     65    75%
release/ray_release/job_manager/__init__.py                       3      3     0%
release/ray_release/job_manager/anyscale_job_manager.py         182    182     0%
release/ray_release/job_manager/job_manager.py                   70     64     9%
release/ray_release/log_aggregator.py                            57     57     0%
release/ray_release/logger.py                                    11     11     0%
release/ray_release/reporter/__init__.py                          0      0   100%
release/ray_release/reporter/artifacts.py                        31     31     0%
release/ray_release/reporter/db.py                               21     21     0%
release/ray_release/reporter/log.py                              21     21     0%
release/ray_release/reporter/reporter.py                         11      7    36%
release/ray_release/result.py                                    87     62    29%
release/ray_release/scripts/__init__.py                           0      0   100%
release/ray_release/scripts/build_pipeline.py                   110    110     0%
release/ray_release/scripts/get_aws_instance_information.py      25     25     0%
release/ray_release/scripts/get_test_summary.py                  23     23     0%
release/ray_release/scripts/ray_bisect.py                        95     95     0%
release/ray_release/scripts/run_release_test.py                  61     61     0%
release/ray_release/signal_handling.py                           33     23    30%
release/ray_release/template.py                                  82      3    96%
release/ray_release/tests/__init__.py                             0      0   100%
release/ray_release/tests/_test_catch_args.py                     9      0   100%
release/ray_release/tests/_test_run_release_test_sh.py           27      0   100%
release/ray_release/tests/test_alerts.py                         18      1    94%
release/ray_release/tests/test_anyscale_job_manager.py           15     15     0%
release/ray_release/tests/test_anyscale_job_wrapper.py           55      1    98%
release/ray_release/tests/test_bisect.py                         30     30     0%
release/ray_release/tests/test_buildkite.py                     293      4    99%
release/ray_release/tests/test_cluster_manager.py               323     16    95%
release/ray_release/tests/test_config.py                         75      1    99%
release/ray_release/tests/test_env.py                            25      2    92%
release/ray_release/tests/test_glue.py                          367      7    98%
release/ray_release/tests/test_log_aggregator.py                 13     13     0%
release/ray_release/tests/test_result.py                         12     12     0%
release/ray_release/tests/test_run_script.py                     53      2    96%
release/ray_release/tests/test_template.py                       33     33     0%
release/ray_release/tests/test_wheels.py                        132      1    99%
release/ray_release/tests/utils.py                               35     16    54%
release/ray_release/util.py                                      99     71    28%
release/ray_release/wheels.py                                   205     93    55%
release/setup.py                                                  2      2     0%
---------------------------------------------------------------------------------
TOTAL                                                          4775   2281    52%
```

### CODE REVIEW

**Review by aslonnie — COMMENTED** [review]
synced offline; please run tests with bazel.

asking to figure a way to get coverage data from running bazel tests .

**Comment by can-anyscale** [comment]
Let's redo this with bazel!

**Comment by can-anyscale** [comment]
Script tuns happily locally as well as on buildkite: https://buildkite.com/ray-project/oss-ci-coverage/builds/29

File: ci/test/ray_coverage.py [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+    coverage DB per test run. Also use 'test' dynamic context so we can store
+    file coverage information.
+    """
+    source_dir = os.path.join(os.getcwd(), "release")
+    subprocess.check_output(
+        [
+            "bazel",
+            "test",
+            test_target,
+            "--test_tag_filters=release_unit",
```
Comment (krfricke) [inline ci/test/ray_coverage.py:40]: Should this filter be dynamic?

File: ci/test/ray_coverage.py [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+
+COVERAGE_FILE_NAME = "ray_release.cov"
+
+
+@click.command()
+@click.argument("test_target", required=True, type=str)
+def main(test_target: str) -> None:
+    """
+    This script collects dynamic coverage data for the test target, and upload the
+    results to database (S3).
```
Comment (krfricke) [inline ci/test/ray_coverage.py:16]: We're not uploading anything to s3 now, right?

**Review by krfricke — COMMENTED** [review]
Looks good!

Should we move the script to `ci/lint` directory and run it in the linter job? We can upload the results as an artifact in buildkite.

File: ci/test/ray_coverage.py (hunk 1 above)
Comment (can-anyscale) [reply ci/test/ray_coverage.py:40]: Eventually it should be yes. I want to do this for release unit tests first though as a prototype and expand it further later.

Comment (can-anyscale) [reply ci/test/ray_coverage.py:40]: Eventually it should be yes. I want to do this for release unit tests first though as a prototype and expand it further later.

File: ci/test/ray_coverage.py (hunk 2 above)
Comment (can-anyscale) [reply ci/test/ray_coverage.py:16]: preemptive comments ;)

**Comment by can-anyscale** [comment]
I plan to run this in a cronjob in buildkite once a day or something along that line. It runs pretty slow so I probably won't put it in linter ;).

**Comment by krfricke** [comment]
> I plan to run this in a cronjob in buildkite once a day or something along that line to collect coverage data. It runs pretty slow so I probably won't put it in linter ;).

Ok I see.

Let's stil ltry to find a better naming for the directory - `ci/test` is too ambiguous (everything is test-related in CI...). How about `ci/metrics`?

**Comment by can-anyscale** [comment]
I know right, feel like everything should be name ci/ci/ci ;).

I notice that we have a ci/pipeline folder. Probably I will reuse that, then I can continue to reuse that later on when I write another script to read this coverage info and determine what tests to run. Hoes does that sound?

File: ci/pipeline/ray_coverage.py [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+
+
+@click.command()
+@click.argument("test_target", required=True, type=str)
+def main(test_target: str) -> None:
+    """
+    This script collects dynamic coverage data for the test target, and upload the
+    results to database (S3).
+    """
+    logger = _get_logger()
```
Comment (aslonnie) [inline ci/pipeline/ray_coverage.py:18]: get logger out in global? or is inlined logger creation a common patter for command line?

File: ci/pipeline/ray_coverage.py [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+    logger.info(f"Collecting coverage for test target: {test_target}")
+    coverage_file = os.path.join(tempfile.gettempdir(), COVERAGE_FILE_NAME)
```
Comment (aslonnie) [inline ci/pipeline/ray_coverage.py:20]: is there going to be anything the deletes this tempdir?

File: ci/pipeline/ray_coverage.py [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+    subprocess.check_output(
+        [
+            "bazel",
+            "test",
+            test_target,
+            "--test_tag_filters=release_unit",
+            "--jobs",
+            "1",
+            "--test_env="
+            f"PYTEST_ADDOPTS=--cov-context=test --cov={source_dir} --cov-append",
```
Comment (aslonnie) [inline ci/pipeline/ray_coverage.py:44]: either make this the same line and use an explicit `+`, or drop the `=` on the previous line and make them two args.

File: ci/pipeline/ray_coverage.py [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+
+def _run_test(test_target: str, coverage_file: str) -> None:
+    """
+    Run test target serially using bazel and compute coverage data using pycov.
+    We need to run tests serially to avoid data races when pytest creates the initial
+    coverage DB per test run. Also use 'test' dynamic context so we can store
+    file coverage information.
+    """
+    source_dir = os.path.join(os.getcwd(), "release")
+    subprocess.check_output(
```
Comment (aslonnie) [inline ci/pipeline/ray_coverage.py:35]: just use `subprocess.check()` as no one is reading the output?

File: ci/pipeline/ray_coverage.py [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+
+
+def _run_test(test_target: str, coverage_file: str) -> None:
+    """
+    Run test target serially using bazel and compute coverage data using pycov.
+    We need to run tests serially to avoid data races when pytest creates the initial
+    coverage DB per test run. Also use 'test' dynamic context so we can store
+    file coverage information.
+    """
+    source_dir = os.path.join(os.getcwd(), "release")
```
Comment (aslonnie) [inline ci/pipeline/ray_coverage.py:34]: so this assumes that the command runs at the source root?

File: ci/pipeline/ray_coverage.py [hunk 8]
```diff
@@ -0,0 +1,69 @@
+import click
```
Comment (aslonnie) [inline ci/pipeline/ray_coverage.py:1]: put click in a different import section.

normally stdlib imports and non-std lib imports are in different sections.

File: ci/pipeline/ray_coverage.py [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+import logging
+import os
+import subprocess
+import sys
+import tempfile
+
+COVERAGE_FILE_NAME = "ray_release.cov"
```
Comment (aslonnie) [inline ci/pipeline/ray_coverage.py:8]: make this private, rename to `_COVERATE_FILE_NAME`?

**Comment by aslonnie** [comment]
not sure if we want to convert these into `py_binary`.. maybe another day..

File: ci/pipeline/ray_coverage.py (hunk 4 above)
Comment (can-anyscale) [reply ci/pipeline/ray_coverage.py:20]: Good point, I think I should use /tmp/artifacts which are setup with auto clean-up

File: ci/pipeline/ray_coverage.py (hunk 7 above)
Comment (can-anyscale) [reply ci/pipeline/ray_coverage.py:34]: good observation, yes, either that or I can make this a command line argument as well

File: ci/pipeline/ray_coverage.py (hunk 3 above)
Comment (can-anyscale) [reply ci/pipeline/ray_coverage.py:18]: get logger out in global sounds good!

**Review by aslonnie — APPROVED** [review]

File: release/requirements_buildkite.txt [hunk 10]
```diff
@@ -1,5 +1,5 @@
 #
-# This file is autogenerated by pip-compile with python 3.7
+# This file is autogenerated by pip-compile with python 3.10
```
[the remark this answers is not in the document]
Comment (aslonnie) [reply release/requirements_buildkite.txt:2]: I am filing a PR to re-resolve the deps in 3.9:

https://github.com/ray-project/ray/pull/35176

[the remark this answers is not in the document]
Comment (can-anyscale) [reply release/requirements_buildkite.txt:2]: Probably because I haven't rebase to master

[the remark this answers is not in the document]
Comment (aslonnie) [reply release/requirements_buildkite.txt]: I tried, I think this update rule will always use the python from the system here.

[the remark this answers is not in the document]
Comment (can-anyscale) [reply release/requirements_buildkite.txt]: Ok, I just updated it to 3.9

**Review by krfricke — APPROVED** [review]
Code lgtm, let's resolve conflicts and ping me to merge

**Comment by ollie-iterators** [comment]
The last failing test is probably unrelated because it is passing on the main branch

**Comment by can-anyscale** [comment]
save

**Comment by can-anyscale** [comment]
hi

**Comment by can-anyscale** [comment]
three

**Pull request closed without merging**